ggml/src/ggml-cuda/fattn-vec-f32.cuh

```diff
@@ -1,7 +1,7 @@
 #include "common.cuh"
 #include "fattn-common.cuh"
 
-template<int D, int ncols, int parallel_blocks, ggml_type type_K, ggml_type type_V> // D == head size
+template<int D, int ncols, int parallel_blocks, ggml_type type_K, ggml_type type_V, bool use_logit_softcap> // D == head size
```

Comment:
Would it affect performance significantly to do this check at runtime? Otherwise this is going to double the number of template instances.

Reply:
When evaluating 8192 tokens with a batch size of 1 on my 3090 a runtime check reduces the end-to-end performance by ~19%. For a batch size of 512 the difference is ~1%. But the latter is only because the kernels using tensor cores are poorly written; due to the use of the WMMA interface data has to be written to and read from shared memory which prevents the compiler from properly reordering instructions (so it doesn't matter if a conditional statement for logit softcapping also does this). My plans are to re-write the kernels as part of my work on the LLM backwards pass in the coming months, I expect the difference for large batches to then be similar to the difference for batch size 1.

Reply:
Anyways, regarding how to move forward: the compilation time/binary size does not grow by a factor of 2 from this PR because you only need the template instances for a head size of 128 or 256, the rest can be skipped. However, because larger head sizes contribute a disproportionate amount to the compilation time/binary size the impact of FlashAttention would still grow by a factor of ~1.7. If that is judged to be too much considering right now there is effectively only a single model that would benefit, I would suggest making Gemma 2 support opt-in in a similar way to `GGML_CUDA_FA_ALL_QUANTS`.

We should then also think about how to handle opt-in features in general. It's basically guaranteed that code for training would add further to compilation time/binary size though most users will not need it. But that case can be covered relatively easily (for llama.cpp) by building the training code only when building at least one example that uses it. Still, for ggml we should maybe add one define per optional CUDA feature and then also a single define like `GGML_CUDA_FULL` that enables all optional features for convenience.

Reply:
I am not concerned about the binary size, but it would be nice to have some option to accelerate the compilation time. I think the simplest way to do that would be to build everything by default, and add options to disable some of the functionality (make it opt-out, rather than having to enable every feature or add a `GGML_CUDA_FULL`). I don't think it is a big issue at the moment, and I don't think we need to make Gemma 2 support opt-in.

Reply:
@slaren Can this be merged? I would very much like the performance improvements through Flash Attention support for the Gemma 2 model.

Reply:
The problem here is not so much slaren but rather me. Basically the only limiting factor for me is how much time I have so I'm trying to optimize throughput rather than latency when it comes to features. My impression was that working on other things first will result in less work for overall so I've put this PR on the backburner.